nannou_wgpu/src/texture/image.rs

```diff
@@ -311,7 +312,14 @@ impl wgpu::RowPaddedBuffer {
     /// polled. You should *not* rely on the being ready immediately.
     pub async fn read<'b>(&'b self) -> Result<ImageReadMapping<'b>, wgpu::BufferAsyncError> {
         let slice = self.buffer.slice(..);
-        slice.map_async(wgpu::MapMode::Read).await?;
+        let (tx, rx) = futures::channel::oneshot::channel();
+
+        slice.map_async(wgpu::MapMode::Read, |res| {
+            tx.send(res).expect("Failed to send map_async result");
+        });
+
+        rx.await.expect("Failed to receive map_async result")?;
+
```

Comment:
I'm currently working on fixing everything in this comment here https://github.com/nannou-org/nannou/pull/940#issuecomment-1762594523

I was able to fix two things, the commits can be found here

- https://github.com/RobWalt/nannou/commit/648668efdb8b6916c5725121712fa22c26ef44d2
- https://github.com/RobWalt/nannou/commit/9a61ce7943147f08858c1443737cb44bee07d7dc

Now wgpu is complaining a lot about these lines here. The docs of `map_async` state

> Map the buffer. Buffer is ready to map once the callback is called.
>
>For the callback to complete, either `queue.submit(..)`, `instance.poll_all(..)`, or `device.poll(..)`
must be called elsewhere in the runtime, possibly integrated into an event loop or run on a separate thread.
>
>The callback will be called on the thread that first calls the above functions after the gpu work
has completed. There are no restrictions on the code you can run in the callback, however on native the
call to the function will not complete until the callback returns, so prefer keeping callbacks short
and used to set flags, send messages, etc.

and I think we don't use any of the 3 mentioned methods here. Not sure how to proceed at the moment

Reply:
Hmm, okay, good catch. This was a breaking change on the upgrade and may have just been an oversight by me. Previously they returned a future from this `map_async` method but claimed the overhead of polling was too high, hence the callback based version that only fires once. I'm not super familiar with this portion of the render pipeline, but might poke around later tonight to see.

Reply:
As far as I'm understanding this, we really have to call one of the three methods

> `queue.submit(..)`, `instance.poll_all(..)`, or `device.poll(..)`

to make that work before awaiting the rx (since the rx is one shot). This is an issue because with the current design it's at best very unclear if this is called at all and at worst isn't called yet. For me this seems like it would need a bigger redesign 😬

Reply:
We flush the queue when the frame is dropped. There were some additional issues with the shaders for this, but the example is running for me now just fine.